### PR TITLE
Rewrite the AST logic to be more complete

### PR DESCRIPTION
**Pull request by yoannchaudet**
Rewrite most of the AST logic to streamline and remove duplicated code.

Notable changes:

(1) We now support all these initializations:

```
Supported configuration initializations:

(1) Default export:

  export default {
    // configuration object here
  }

(2) Direct module export:

  module.exports = {
    // configuration object here
  }

(3) Indirect module export:

  const config = // configuration object here
  module.exports = config
```

REVIEW NOTE: All the meaningful changes are in the `config-parser.js` file.

What else has changed:

- More tests (low level for the AST, high level for the actual configuration injection)
- For next we do a bit more meddling because we have to... trailing slashes are removed on the path, and we also set https://nextjs.org/docs/api-reference/next/image#unoptimized to true (so the configuration injection has been refactored to handle arbitrary property injection)

---

Validation with most basic getting started guides:

- next: ✅ https://github.com/yoannchaudet/nextjs-npx
- nuxt: ✅ https://yoannchaudet.github.io/nuxt/
- gatsby: ✅ https://yoannchaudet.github.io/gatsby/

### CODE REVIEW

File: src/config-parser.js [hunk 1]
```diff
@@ -1,205 +1,281 @@
...
+      // Matching this repo's prettier configuration
+      printWidth: 80,
+      tabWidth: 2,
+      useTabs: false,
+      semi: false,
+      singleQuote: true,
+      trailingComma: 'none',
+      bracketSpacing: false,
+      arrowParens: 'avoid'
+    })
```
Comment (yoannchaudet) [inline src/config-parser.js:270]: I'll scrap that

Comment (JamesMGreene) [reply src/config-parser.js:270]: Yeah, seems like unnecessary overhead here. JavaScript doesn't care about indentation, so we can get away with sloppy formatting for these transient deployment-only files. 🤷🏻‍♂️

File: src/fixtures/expected/gatsby-config.js [hunk 2]
```diff
@@ -1,8 +1,8 @@
-import { resolve } from 'path'
+import {resolve} from 'path'
```
Comment (JamesMGreene) [inline src/fixtures/expected/gatsby-config.js:1]: I'm cringing. 😬 😅

File: src/fixtures/expected/nuxt.config.missing.js [hunk 3]
```diff
@@ -1,11 +1,9 @@
-import { resolve } from 'path'
+import {resolve} from 'path'
 
 export default {
-  router: {
-        base: '/amazing-new-repo/'
-    },
-alias: {
-    'style': resolve(__dirname, './assets/style')
+  router: {base: '/amazing-new-repo/'},
```
Comment (JamesMGreene) [inline src/fixtures/expected/nuxt.config.missing.js:4]: 🙈 😅

File: src/fixtures/x.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  pathPrefix: "hello",
```
Comment (JamesMGreene) [inline src/fixtures/x.js:2]: More consistent styling:

```suggestion
  pathPrefix: 'hello'
```

File: src/set-pages-path.js [hunk 5]
```diff
@@ -1,41 +1,38 @@
 const core = require('@actions/core')
-const axios = require('axios')
 const {ConfigParser} = require('./config-parser')
 
+function getParserConfiguration(staticSiteGenerator, path) {
+  switch (staticSiteGenerator) {
+    case 'nuxt':
+      return {
+        configurationFile: './nuxt.config.js',
+        propertyName: 'router.base',
```
Comment (JamesMGreene) [inline src/set-pages-path.js:9]: Glad to see this switch to arbitrary path depths. ✨

File: src/set-pages-path.js [hunk 6]
```diff
@@ -1,41 +1,38 @@
...
+        propertyValue: path,
+        blankConfigurationFile: `${process.cwd()}/blank-configurations/nuxt.js`
```
Comment (JamesMGreene) [inline src/set-pages-path.js:11]: :warning: Seems like we are missing a `src/` or `dist/` path segment here, but we could avoid that by using `__dirname` instead. 🤔 

```suggestion
        blankConfigurationFile: `${__dirname}/blank-configurations/nuxt.js`
```

**Review by JamesMGreene — COMMENTED** [review]
I like the overall capability increase that these changes appear to be granting us. It does seem like there are some minor issues (e.g. missing path segments), but I suspect those will proven/shaken out when adding more tests tomorrow. 🤞🏻

File: src/config-parser.js [hunk 7]
```diff
@@ -1,205 +1,281 @@
...
+          node.type === 'VariableDeclaration' &&
+          node.declarations.length == 1 &&
+          node.declarations[0].type === 'VariableDeclarator' &&
+          node.declarations[0].id.type === 'Identifier' &&
+          node.declarations[0].id.name === identifierName &&
+          node.declarations[0].init.type === 'ObjectExpression'
+      )
+      if (identifierDefinition) {
+        core.info('Found configuration object in indirect module export')
+        return identifierDefinition.declarations[0].init
```
Comment (JamesMGreene) [inline src/config-parser.js:118]: When Nuxt 3.x is eventually GA (currently a release candidate), the expected configuration file syntax will change to include a function invocation wrapping the configuration object as well. 😓 

I don't _think_ this code will cover that eventuality...? 🤔 

```js
import { defineNuxtConfig } from 'nuxt'

export default defineNuxtConfig({
  // ...
})
```

From: https://v3.nuxtjs.org/migration/configuration#migration

File: src/fixtures/x.js (hunk 4 above)
Comment (yoannchaudet) [reply src/fixtures/x.js:2]: That's my secret "TDD" fixture. It got check in by accident 🙃

**Comment by JamesMGreene** [comment]
Confirmed that this branch does not yet fix the Next.js starter workflow test repo's verification:

Before, using `configure-pages@main`:

https://github.com/paper-spa/starter-test-nextjs/runs/7414265302?check_suite_focus=true

```
> Build error occurred
/home/runner/work/starter-test-nextjs/starter-test-nextjs/next.config.js:1
export default {
^^^^^^
SyntaxError: Unexpected token 'export'
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1033:15)
    at Module._compile (node:internal/modules/cjs/loader:106[9](https://github.com/paper-spa/starter-test-nextjs/runs/7414265302?check_suite_focus=true#step:7:10):27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:[10](https://github.com/paper-spa/starter-test-nextjs/runs/7414265302?check_suite_focus=true#step:7:11))
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:[12](https://github.com/paper-spa/starter-test-nextjs/runs/7414265302?check_suite_focus=true#step:7:13))
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:[17](https://github.com/paper-spa/starter-test-nextjs/runs/7414265302?check_suite_focus=true#step:7:18)0:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:[19](https://github.com/paper-spa/starter-test-nextjs/runs/7414265302?check_suite_focus=true#step:7:20)8:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:385:24)
Error: Process completed with exit code 1.
```

After, using `configure-pages@fallback`:

https://github.com/paper-spa/starter-test-nextjs/runs/7414838915?check_suite_focus=true

_(Same error.)_

---

The test repo currently has no `next.config.js`. Looks like we are defaulting to using an ESM syntax base configuration if that is the case, which AFAIK is not actually supported by Next.js without additional configuration/build.

To fix it, we could:
- default to writing the file in CommonJS syntax
  - 💚 **Recommended** [by me]
- write the file out as `next.config.mjs` (**NOTE:** `.mjs` extension)
  - This gets more complicated... because they we should also be checking for the existence of the `next.config.mjs` file as well, which adds more layers of complexity
- Check the `package.json` file for its `type` property to see if it is `type: "module"` and we _should_ write it out as an ESM syntax; otherwise, write it out as a CommonJS syntax
  - Again, this adds more layers of complexity

**Comment by JamesMGreene** [comment]
**Correction:** This PR remedies that problem already. ☝🏻 🎉  I just pushed the updated `dist/` files. 😉 

---

Still running into other errors, though. :sob:

https://github.com/paper-spa/starter-test-nextjs/runs/7415147409?check_suite_focus=true

https://nextjs.org/docs/messages/export-image-api

**Comment by JamesMGreene** [comment]
```
Run npx --no-install next export
info  - using build directory: /home/runner/work/starter-test-nextjs/starter-test-nextjs/.next
info  - Copying "static build" directory
info  - No "exportPathMap" found in "undefined". Generating map from "./pages"
Error: Image Optimization using Next.js' default loader is not compatible with `next export`.
  Possible solutions:
    - Use `next start` to run a server, which includes the Image Optimization API.
    - Configure `images.unoptimized = true` in `next.config.js` to disable the Image Optimization API.
  Read more: https://nextjs.org/docs/messages/export-image-api
    at /home/runner/work/starter-test-nextjs/starter-test-nextjs/node_modules/next/dist/export/index.js:1[5](https://github.com/paper-spa/starter-test-nextjs/runs/7415147409?check_suite_focus=true#step:8:6)0:23
    at async Span.traceAsyncFn (/home/runner/work/starter-test-nextjs/starter-test-nextjs/node_modules/next/dist/trace/trace.js:[7](https://github.com/paper-spa/starter-test-nextjs/runs/7415147409?check_suite_focus=true#step:8:8)[9](https://github.com/paper-spa/starter-test-nextjs/runs/7415147409?check_suite_focus=true#step:8:10):20)
Error: Process completed with exit code 1.
```

**Comment by yoannchaudet** [comment]
@JamesMGreene I almost got a heart attack seeing your message :trollface: 

It's still a big mess in the PR, I added `set-pages-path` tests. I am going to bring back low level tests for the actual AST in next commits.

Also the distribution while updated is still not correct. I need it to include the blank templates for gatsby/next/nuxt. They are not in the right place now.

File: src/config-parser.js (hunk 7 above)
Comment (yoannchaudet) [reply src/config-parser.js:117]: Not yet but that should be simple to add with this re-write.

**Comment by yoannchaudet** [comment]
Scrap what I said about the distribution being incorrect, `ncc` is actually smart enough to rewrite the paths 🤯 That's great.

**Pull request merged**